src/util/cache.cpp

```diff
@@ -0,0 +1,31 @@
+#include "util/cache.h"
+
+#include "util/assert.h"
+
+namespace mixxx {
+
+cache_key_t calculateCacheKey(const QByteArray& bytes) {
```

Comment:
Can you explain what the rationale for this function is? If the goal is to save the hash in a quint64, why don't you just truncate it?

Reply:
Truncation would imply that all bytes in the array have the same significance. The less assumptions we need to make, the more versatile the function is. Combining them by XOR is a cheap operation, we don't need to care. I will add a comment.

Reply:
> Truncation would imply that all bytes in the array have the same significance.

Due to the nature of cryptographic hashes and avalanche effect (changing a single input bit should result in change of at least half of the output bits), they should.

According to the NIST, truncation is the official way to deal with situations like these:

> ## 7.              TRUNCATION OF A MESSAGE DIGEST 
> Some application may require a hash function with a message digest length different than those provided by the hash functions in this Standard. In such cases, a truncated message digest may be used, whereby a hash function with a larger message digest length is applied to the data to be hashed, and the resulting message digest is truncated by selecting an appropriate number of the leftmost bits. For guidelines on choosing the length of the truncated message digest and information about its security implications for the cryptographic application that uses it, see SP 800-107 [SP 800-107]. 
> *Source:* [FIPS 180-4 Secure Hash Standard (SHS)](https://csrc.nist.gov/publications/detail/fips/180/4/final)

And from SP 800-107:
> ## 5 Hash function Usage 
> ### 5.1      Truncated Message Digest 
> [...]
> 1. The length of the output block of the approved hash function to be used shall be greater than λ (i.e., L > λ). 
> 2. The λ left-most bits of the full-length message digest shall be selected as the 
truncated message digest. 
> For example, if a truncated message digest of 96 bits is desired, the SHA-256 
hash function could be used (e.g., because it is available to the application, and 
provides an output larger than 96 bits). The leftmost 96 bits of the 256-bit 
message digest generated by SHA-256 are selected as the truncated message 
digest, and the rightmost 160 bits of the message digest are discarded.  
> 3. If collision resistance is required, λ shall be at least twice the required collision 
resistance strength s (in bits) for the truncated message digest (i.e., λ ≥ 2s). 
> [...]
> *Source:* [SP 800-107 Rev. 1 Recommendation for Applications Using Approved Hash Algorithms](https://csrc.nist.gov/publications/detail/sp/800-107/rev-1/final)

So for λ=64 bits we'd get a collision resistance of 2³² (Preimage resistance and second preimage resistance are of not really a concern here).